### PR TITLE
MD-1954 - Add support to run tests tagged as failures

### PR DESCRIPTION
**Pull request by kedarbcs16**
Default for .fail extension. 
To override, provide fail-extension in test definition file

### CODE REVIEW

**Comment by rhou1** [comment]
Can you comment on what you have done?  Have you tested the code.  If so, what testing have you done?

**Comment by kedarbcs16** [comment]
Hi Robert 

Yes I have tested the code on Functional/joins/nulleqjoin  and Functional/cttas as of now .
The changes were 
I changed the regex to pick up default (.fail and .failing) or any other format specified in the def file in fail-extension key instead of only .q or .sql file if the cmdline argument z is set to true .

Thanks

Kedar Sankar Behera

**Comment by rhou1** [comment]
The code is a little longer than I expected.  I need to spend some time to understand it.  And we are missing the report of tests that are passing now.

Also, I think there are failed tests that are not marked with "fail" or "failing".  We should check for "fail" in query filenames.

**Comment by Agirish** [comment]
Kedar, please squash your commits.

File: framework/src/main/java/org/apache/drill/test/framework/CmdParam.java [hunk 1]
```diff
@@ -52,6 +52,9 @@
     @Parameter(names = {"-r", "--report"}, description = "Generate json report", required=false)
     public boolean generateReports = false;
 
+    @Parameter(names = {"-z", "--negativeCases"}, description = "Include fail test cases", required=false)
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/CmdParam.java]: Change name to "-if" --"includeFailed"
Change description to "Include tests tagged as failures"

File: framework/src/main/java/org/apache/drill/test/framework/CmdParam.java [hunk 2]
```diff
@@ -52,6 +52,9 @@
...
+    public boolean negativeCasesInclusion = false;
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/CmdParam.java]: These are not negative tests. Change variable name to "includeFailed"

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 3]
```diff
@@ -226,33 +241,75 @@ public int runTests() throws Exception {
...
+
+      if(TestDriver.cmdParam.negativeCasesInclusion == true){
+      	LOG.info(LINE_BREAK);
+	int temp = 0;
+      	for(DrillTest test : passingTests){
+      		//String inputFile = test.getInputFile();
+		//String failExtension = modeler.matrices.get(0).failExtension;
+		//if(failExtension==null){
+			 //LOG.info("blej "+test.modeler.matrices.get(0).schema);
+			 //LOG.info("blej bleh b;leh ");
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Remove not needed lines, instead of commenting out.

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 4]
```diff
@@ -226,33 +241,75 @@ public int runTests() throws Exception {
...
       LOG.info(LINE_BREAK);
+      //TestCaseModeler modeler;
+      //  try {
+      //    modeler = getTestCaseModeler(testDefFile.getAbsolutePath());
+      //  } catch (JsonParseException e) {
+      //    LOG.warn("Caught exception parsing " + testDefFile + ". This test will not be executed.", e);
+      //    continue;
+      //  }
+
+      if(TestDriver.cmdParam.negativeCasesInclusion == true){
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Add comments to give a brief description of the change.

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 5]
```diff
@@ -226,33 +241,75 @@ public int runTests() throws Exception {
...
+				}
+				temp++;
+				LOG.info(test.getInputFile());
+			}
+		//}
+		//else{
+		//	if(inputFail.contains(failExtension)){
+		//		LOG.info(test.getInputFile());
+		//	}	
+		//}
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Again, please remove these lines

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java [hunk 6]
```diff
@@ -167,9 +168,25 @@
...
+        expectedFileExtension = FilenameUtils.getExtension(expectedFileExtension);
+ 
+        String failExtension = modeler.matrices.get(0).failExtension;
+        String tempQueryExt = queryFileExtension;
+        if(TestDriver.cmdParam.negativeCasesInclusion == true){
+            String fileExt = FilenameUtils.getExtension(queryFileExtension);
+	    String baseExt = FilenameUtils.removeExtension(queryFileExtension);
+	    if(failExtension!=null){
+		failExtension = FilenameUtils.getExtension(failExtension);
+	    	queryFileExtension = baseExt+".("+fileExt+"|"+failExtension+")";
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: Does this handle cases where there is no base extension?

Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: yes

File: framework/src/main/java/org/apache/drill/test/framework/CmdParam.java (hunk 1 above)
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/CmdParam.java]: done

File: framework/src/main/java/org/apache/drill/test/framework/CmdParam.java (hunk 2 above)
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/CmdParam.java]: done

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java (hunk 4 above)
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: done

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java (hunk 3 above)
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: done

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java (hunk 5 above)
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: done

File: framework/src/main/java/org/apache/drill/test/framework/CmdParam.java [hunk 7]
```diff
@@ -52,10 +52,13 @@
     @Parameter(names = {"-r", "--report"}, description = "Generate json report", required=false)
     public boolean generateReports = false;
 
+    @Parameter(names = {"-if", "--includeFailed"}, description = "Include tests tagged as failures", required=false)
```
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/CmdParam.java]: cmd line argument to include test tagged as failure

File: framework/src/main/java/org/apache/drill/test/framework/TestCaseModeler.java [hunk 8]
```diff
@@ -86,6 +86,8 @@ public static TestCaseModeler createTestCase(String testId, String type,
     public String outputFormat;
     @JsonProperty("expected-file")
     public String expectedFile;
+    @JsonProperty("fail-extension")
+    public String failExtension;
```
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/TestCaseModeler.java]: fail extension from the test definition json

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 9]
```diff
@@ -188,10 +190,21 @@ public int runTests() throws Exception {
       List<DrillTest> executionFailures = Lists.newArrayList();
       List<DrillTest> timeoutFailures = Lists.newArrayList();
       List<DrillTest> canceledTests = Lists.newArrayList();
+      List<DrillTest> failedCases = Lists.newArrayList();
 
       for (DrillTest test : tests) {
         TestStatus testStatus = test.getTestStatus();
-        switch (testStatus) {
+	if(cmdParam.includeFailed==true){
```
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: get the list of test cases tagged as failures

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 10]
```diff
@@ -226,33 +241,57 @@ public int runTests() throws Exception {
...
       }
-      LOG.info("Timeout Failures:");
+      if(timeoutFailures.size()>0){
+      	LOG.info("Timeout Failures:");
+      }
       for (DrillTest test : timeoutFailures) {
         LOG.info(test.getInputFile());
       }
       LOG.info(LINE_BREAK);
+
```
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: log the tests tagged as failure that passed

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 11]
```diff
@@ -265,16 +304,22 @@ public int runTests() throws Exception {
       }
 
       totalPassingTest += passingTests.size();
+      for(DrillTest test : passingTests){
```
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: final Count of test cases tagged as failures that passed

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java [hunk 12]
```diff
@@ -167,9 +168,25 @@
           }
         }
         if (!foundTests) {continue;}
-        
-        String queryFileExtension = modeler.matrices.get(0).inputFile;
+       
+	String queryFileExtension = modeler.matrices.get(0).inputFile;
         String expectedFileExtension = modeler.matrices.get(0).expectedFile;
+
```
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: changed the regex string to include .fail/.failed or any other failextension provided in the testDefFile

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java [hunk 13]
```diff
@@ -220,11 +237,35 @@ private static TestCaseModeler getTestCaseModeler(String testDefFile)
...
+	    return queryFile;
+
+  }
+
   
   private static String getExpectedFile(String queryFile, String queryFileExt,
 	      String expectedFileExt) {
-	    int idx = queryFile.indexOf(queryFileExt.substring(2));
-	    return queryFile.substring(0, idx).concat(expectedFileExt.substring(2));
+
```
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: Changed the getExpectedFile function as it was not supporting the test files that didnt have a .q or .sql in its filename .

File: framework/src/main/java/org/apache/drill/test/framework/CmdParam.java (hunk 7 above)
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/CmdParam.java]: now that we are only going to run failed tests, may be we should rename this to -rf, --runFailed. Also change description

**Comment by Agirish** [comment]
Please rebase on mapr/master, squash your commits and address the minor review comments.

**Comment by kedarbcs16** [comment]
Jenkins job - http://10.10.104.91:8080/view/MD-1954/job/MD-1954/9/consoleFull

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java [hunk 14]
```diff
@@ -167,8 +168,28 @@
           }
         }
         if (!foundTests) {continue;}
+
         String queryFileExtension = modeler.matrices.get(0).inputFile;
+
         String expectedFileExtension = modeler.matrices.get(0).expectedFile;
+
+        expectedFileExtension = FilenameUtils.getExtension(expectedFileExtension);
```
Comment (rhou1) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: Is this line needed?

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java [hunk 15]
```diff
@@ -167,8 +168,28 @@
...
+ 
+        String failExtension = modeler.matrices.get(0).failExtension;
+        String tempQueryExt = queryFileExtension;
```
Comment (rhou1) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: Is this line needed?

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java [hunk 16]
```diff
@@ -167,8 +168,28 @@
...
+        //To include fail extension in the regex to pick query files tagged as failure as well 
+	if(TestDriver.cmdParam.runFailed == true){
+            String fileExt = FilenameUtils.getExtension(queryFileExtension);
+	    String baseExt = FilenameUtils.removeExtension(queryFileExtension);
```
Comment (rhou1) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: Is this line needed?

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java [hunk 17]
```diff
@@ -219,11 +242,39 @@ private static TestCaseModeler getTestCaseModeler(String testDefFile)
 	    ObjectMapper objectMapper = new ObjectMapper();
 	    return objectMapper.readValue(new String(jsonData), TestCaseModeler.class);
 	  }
+
+  /**
+   * @return returns the base name of a file with or without multiple extensions
+   *         
+   */  
+  private static String getBaseName(String queryFile){
```
Comment (rhou1) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: Is this procedure needed?

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java (hunk 13 above)
Comment (rhou1) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: Why can't the original code be used, with perhaps a modification?

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java [hunk 18]
```diff
@@ -181,10 +199,13 @@
...
                   queryFileExtension);
           for (File testQueryFile : testQueryFiles) {
-            String expectedFileName = getExpectedFile(testQueryFile.getAbsolutePath(),
-            		queryFileExtension, expectedFileExtension);
-            drillTestCases.add(new DrillTestCase(modeler, testQueryFile.getAbsolutePath(), expectedFileName));
-          }
+
+	    String expectedFileName = getExpectedFile(testQueryFile.getAbsolutePath(),modeler.matrices.get(0).inputFile,
+                      expectedFileExtension);
+	    if(expectedFileName!=null && expectedFileName!=""){
```
Comment (rhou1) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: Can expectedFileName be null?

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java [hunk 19]
```diff
@@ -219,12 +240,53 @@ private static TestCaseModeler getTestCaseModeler(String testDefFile)
...
+
+	    	int idx = queryFile.indexOf(queryFileExt.substring(2));
+	    	if(idx>=0){
+			return queryFile.substring(0, idx).concat(expectedFileExt.substring(2));
+	    	}
+		else{
+			LOG.info("expectedFile not found for "+queryFile);
+		}
+	    } catch (Exception name) {
+		LOG.info("expectedFile not found for "+queryFile);
```
Comment (rhou1) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: Can we add a comment with an example of when this happens?

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java [hunk 20]
```diff
@@ -181,10 +199,13 @@
...
           for (File testQueryFile : testQueryFiles) {
-            String expectedFileName = getExpectedFile(testQueryFile.getAbsolutePath(),
-            		queryFileExtension, expectedFileExtension);
-            drillTestCases.add(new DrillTestCase(modeler, testQueryFile.getAbsolutePath(), expectedFileName));
-          }
+
+	    String expectedFileName = getExpectedFile(testQueryFile.getAbsolutePath(),modeler.matrices.get(0).inputFile,
+                      expectedFileExtension);
+	    if(expectedFileName!=null && expectedFileName!=""){
+            	drillTestCases.add(new DrillTestCase(modeler, testQueryFile.getAbsolutePath(), expectedFileName));
```
Comment (rhou1) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: What happens if we add the test case if the expectedFileName is blank or null?

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java (hunk 18 above)
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: no i wrote that just as an extra precaution ,should I remove it

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java (hunk 20 above)
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: I dont know this behavior but whats the point to run the test for this ,We can check how it behaves

File: framework/src/main/java/org/apache/drill/test/framework/Utils.java (hunk 19 above)
Comment (kedarbcs16) [inline framework/src/main/java/org/apache/drill/test/framework/Utils.java]: Example :- 

 /root/drill-test-framework/framework/resources/Functional/partition_pruning/dfs/parquet/plan/drill4860.e_tsv.fail

/root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/noRowsLHS.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/noRowsSubQryRHS.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/emptyLHS_RHS_JSON.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/filterFalse_RHS.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/emptyLHS_CSV.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/queryEmptyCSV.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/emptyRHS_JSON.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/queryEmptyJSON.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/emptyLHS_JSON.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/filterFalse_LHS.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/noRowsRHS.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/emptyRHS_CSV.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/emptyLHS_RHS_CSV.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/union_all/prq_union_all/noRowsSubQryLHS.DRILL4476.failing

 /root/drill-test-framework/framework/resources/Functional/joins/nulleqjoin/neqjn_intrvldy.failing

 /root/drill-test-framework/framework/resources/Functional/joins/nulleqjoin/innrjn_intrvlyr.failing

 /root/drill-test-framework/framework/resources/Functional/joins/nulleqjoin/neqjnINsubqry_2.failing

 /root/drill-test-framework/framework/resources/Functional/joins/nulleqjoin/cte_neqjn_union.failing

 /root/drill-test-framework/framework/resources/Functional/joins/nulleqjoin/neqjn_intrvlyr.failing

 /root/drill-test-framework/framework/resources/Functional/joins/nulleqjoin/innrjn_intrvlday.failing

 /root/drill-test-framework/framework/resources/Functional/window_functions/tpcds/query89.failing

 /root/drill-test-framework/framework/resources/Functional/window_functions/tpcds/query57.failing

 /root/drill-test-framework/framework/resources/Functional/window_functions/tpcds/query47.failing
